# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [3]:
# Import your libraries:

import numpy as np
import pandas as pd


# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [4]:
# Your code here:
import sklearn.datasets
diabetes = sklearn.datasets.load_diabetes()


Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [5]:
# Your code here:
diabetes

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990749, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06833155, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286131, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04688253,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452873, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00422151,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [6]:
# Your code here:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

In [7]:
diabetes.data.shape

(442, 10)

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

# Enter your answer here:
1.There are 7 attributes in the data: data, target, DESCR, feature_names, data_filename, target_filename & data_module.
data- contains features(independent variables)of diabetes dataset
target- contains target(dependent variable)
DESCR- contains description of dataset
feature_name-contains names of features(columns)
data_filename- provide path to csv file containing data.
target_filename- provide path to csv file containing target.
data_module- shows data is loaded from sklearn.datasets
2.The relation between diabetes['data'] and diabetes['target'] is that diabetes['data'] contains features(independent variable)
that are used as input in ML model to predict the target that is stored in diabetes['target']. diabetes['target'] contains
the target(dependent variable), that is to be predict by the ML model on the basis of features which we provide to the model.
3.There are 442 records in the data.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [9]:
# Your code here:
number_records_data = diabetes.data.shape
number_records_data


(442, 10)

In [10]:
number_records_target= diabetes.target.shape
number_records_target

(442,)

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [11]:
# Your code here:
from sklearn.linear_model import LinearRegression


#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [12]:
# Your code here:
X = diabetes.data
y = diabetes.target
diabetes_model = LinearRegression()


#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [13]:
# Your code here:
from sklearn.model_selection import train_test_split
diabetes_data_train, diabetes_data_test, diabetes_target_train, diabetes_target_test = train_test_split(X, y, test_size=20, random_state=42)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [14]:
# Your code here:
model = diabetes_model.fit(diabetes_data_train, diabetes_target_train)


In [15]:
print("Intercept:", model.intercept_)
print("Coefficients:", model.coef_)


Intercept: 151.3820926783504
Coefficients: [   0.88135058 -260.26547413  536.46965804  335.9950993  -835.78537996
  465.75983085  136.20013901  245.06432031  712.32537762   64.48680278]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [16]:
# Your code here:
y_pred = model.predict(diabetes_data_test)
print(y_pred)

[141.94619731 176.31674205 134.53159785 287.22526514 122.58042918
  96.46043898 256.8468282  184.79870459  92.90893919 111.3914492
  98.52709941 166.01405738  58.45723061 205.10018437 100.72748232
 130.33505833 218.45993591 245.87432278 193.00193698 214.38106404]


#### Print your `diabetes_target_test` and compare with the prediction. 

In [17]:
# Your code here:
np.array(diabetes_target_test)

array([219.,  70., 202., 230., 111.,  84., 242., 272.,  94.,  96.,  94.,
       252.,  99., 297., 135.,  67., 295., 264., 170., 275.])

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

# Your explanation here:
No, diabetes_target_test is not exactly the same as the model prediction. It means model is not good in predicting 
the actual value of target.

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [19]:
# Your code here:
auto = pd.read_csv('auto-mpg.csv')


Look at the first 5 rows using the `head()` function:

In [20]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [21]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [22]:
# Your code here:
auto['model_year'].unique()
# newest model year is 82 and oldest model year is 70.

array([70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82], dtype=int64)

Check the dataset for missing values and remove all rows containing at least one missing value.

In [23]:
# Your code here:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [24]:
auto.dropna(axis=0,inplace=True)

In [25]:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [26]:
# Your code here:
auto['cylinders'].value_counts()
# There are 5 possible values of cylinders.

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [27]:
# Your code here:
auto.drop('car_name',axis=1,inplace=True)

In [28]:
# X-y split
X = auto.drop('mpg', axis=1)
y = auto['mpg']

In [29]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [30]:
# Your code here:
auto_model = LinearRegression()
model = auto_model.fit(X_train, y_train)

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). In the end, we want the r-squared score to be as high as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [31]:
# Your code here:
y_pred = model.predict(X_train)
print(y_pred)

[20.95609215 26.26333353 22.56255838 12.46318407 34.74115023 14.9315701
 34.11702802 28.24073246 23.96807779 17.1302593  16.80037432 26.61611355
 26.30794398 30.98121298 30.87528248 21.56210797 17.9186156  20.76278751
  9.78482968 27.80721377 20.2273002   9.02193198 16.9392466   9.44761706
 26.94182756 30.89559471 33.62331745 35.48118328 23.83617821 22.99783358
 19.92784153 34.86631962 32.73076784 25.58508818 24.44541041 33.03345258
 29.73872525 19.65564792 21.8665939  25.30949132 21.51340567 30.19618874
 21.0390536  19.72002199 34.94853571 24.67626203 26.00988361 30.27257564
 29.4691753  20.16148151 32.43933728 30.59015328 26.98917691 22.01162485
 32.09549838 17.87589388 27.13479377 25.93193653 32.43810108 17.64219516
 19.01086037 15.98181995 20.09087357 25.34648167 24.58532876 25.92034748
 29.96478919 25.62559361  9.20572947 23.29294559 27.7657253  14.49908127
 30.20582209  8.14795639 23.59746876 31.2493567  25.88569687 32.60824698
 13.77599452 33.58530799 32.97037611  7.7432387  31.

In [32]:
from sklearn.metrics import r2_score
r2_score_value = r2_score(y_train, y_pred)
r2_score_value

0.8107227953093896

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [33]:
# Your code here:
y_test_pred = model.predict(X_test)
print(y_test_pred)
r2_score_value = r2_score(y_test, y_test_pred)
print("r2_score_value of predicted y_test_pred & the oberserved y_test data:",r2_score_value)

[25.93279618 26.29927859 32.96548909 26.8544302  29.45372581 29.08699476
  7.63874768 29.23704446 20.87736024 28.91582943 12.48503673 23.69898277
 16.41889199 28.19668802 22.03967916 30.88152399 21.18311073 31.64662323
 27.91649795 29.78148169 19.84374006 34.37342564 33.99070737 15.02368137
 28.76984129 25.99468114 20.82555389 16.63300722 28.50008896 23.82955469
 12.81866721 23.60612254 21.36279705 29.88377781 11.05575113 34.7003572
 10.84515757 26.26088449 11.58087074  7.71809963 12.77118771 27.81143288
 34.74146456 26.49024951 11.3582456   9.10855962 17.66893597 31.4355561
 25.43286643 30.38576336 11.60450412 25.152572   24.52253515 33.49442837
 28.90462724 17.82504408 20.7761521  23.16942494 23.10923128 24.6273826
  7.683821   22.9296696  27.33070582 23.09519024 28.13950034 28.69699507
 26.85355066 29.61277981 22.23472064  9.37750077 23.83479875 13.39811055
 25.16730444 26.8841077  24.18222144 25.88860198 13.58449896 15.4409455
 24.54031132]
r2_score_value of predicted y_test_pred &

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [34]:
# Your code here:
from sklearn.model_selection import train_test_split
X_train09, X_test09, y_train09, y_test09 = train_test_split(X, y, test_size=0.10, random_state=42)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [35]:
# Your code here:
auto_model09 = LinearRegression()
model= auto_model09.fit(X_train09, y_train09)

Compute the predicted values and r squared score for our new model and new sample data.

In [36]:
# Your code here:
y_pred09 = model.predict(X_test09)
print(y_pred09)


[25.79500879 26.30607444 32.72173147 26.76455357 29.32069319 28.92374612
  7.92293591 29.07543422 20.77292415 28.65066633 12.87802843 23.51587538
 16.49986135 28.03827358 21.94037251 30.75434656 21.04766999 31.44623949
 27.74666775 29.6547018  19.75740051 34.12943268 33.73702057 15.0134863
 28.55382092 25.918776   20.786725   16.59998627 28.29350268 23.75655308
 12.87765797 23.47874091 21.1755328  29.7008905  11.24332729 34.4330599
 10.79531146 26.17242906 11.57563638  8.1727483 ]


In [37]:
r2_score_value = r2_score(y_test09, y_pred09)
print('r2_score_value:',r2_score_value)

r2_score_value: 0.8468911998183242


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [38]:
# Your code here:
X_train_smaller,X_test_smaller, y_train_smaller, y_test_smaller = train_test_split(X,y, test_size=0.30, random_state=42)


In [39]:
auto_model_smaller = LinearRegression()
model= auto_model_smaller.fit(X_train_smaller, y_train_smaller)

In [40]:
y_pred_smaller = model.predict(X_test_smaller)
print(y_pred_smaller)

[26.31050971 26.24667456 33.14488022 27.22440462 29.46550187 29.28389189
  7.66404492 29.42701039 21.1231744  29.36858154 12.23470009 23.80655125
 16.26874891 28.42719194 21.98956832 30.87559458 21.45129619 31.71360155
 28.05590416 29.80098707 20.18737079 34.56036778 34.05061032 15.20319734
 29.06219757 26.19733355 21.20662016 16.96542107 28.97312287 24.19882411
 12.87870775 23.78065106 21.34607145 30.0053973  11.10814518 34.73794534
 11.0471502  26.54534821 11.78380697  7.4650836  13.00060451 28.06576046
 34.71451615 26.75068294 11.43444461  8.71851254 17.98957713 31.57479656
 25.76957248 30.57744654 11.68076577 25.34346015 24.75938746 33.38831804
 28.72102568 18.04557897 20.92253424 23.31121309 23.39290664 24.83692246
  7.25169314 23.14605999 27.23268168 23.36315904 28.16645706 28.8512881
 26.96492867 29.83569432 22.31874769  9.13667285 24.23497322 13.24423311
 25.30451145 27.05157028 24.34658254 26.17427555 13.64025723 15.46768487
 24.8259113  21.07215584 26.38158976 22.77669633 12.

In [41]:
r2_score_value = r2_score(y_test_smaller, y_pred_smaller)
print('r2_score_value:',r2_score_value)

r2_score_value: 0.8076704419810162


conclusion:r2 score for test_size-0.20 is 0.8107,
           r2 score for test_size-0.10 is 0.8468 &
           r2 score for test_size-0.30 is 0.8076. 
This shows that as we decrease the test_size the r2 score decreases. It means that there is improvement in test r squared on decreasing the test size(or increasing training data).

# Bonus Challenge - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [42]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [43]:
# Your code here:
RFE_model = RFE(auto_model,n_features_to_select=3)

Fit the model and print the ranking

In [44]:
# Your code here:
RFE_model.fit(X_train, y_train)
RFE_model.ranking_

array([1, 4, 3, 2, 1, 1])

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [56]:
# Your code here:
# Select the features based on RFE ranking
selected_features = [feature for feature, rank in zip(X_train.columns, RFE_model.ranking_) if rank == 1]
selected_features

['cylinders', 'acceleration', 'model_year']

In [51]:
# Your code here: 
# Create a DataFrame with the selected features
X_reduced = X_train[selected_features]


In [52]:
# Perform a test-train split with reduced feature set
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(
    X_reduced, y_train, test_size=0.2, random_state=42
)

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [53]:
# Create and fit a new linear regression model with reduced features
auto_model_reduced = LinearRegression()
auto_model_reduced.fit(X_train_reduced, y_train_reduced)

LinearRegression()

In [54]:
# Make predictions on the test set
y_pred_reduced = auto_model_reduced.predict(X_test_reduced)


In [55]:
# Calculate R-squared score
r2_score_reduced = r2_score(y_test_reduced, y_pred_reduced)
# Print the R-squared score for the model with reduced features
print(f"R-squared score for the model with reduced features: {r2_score_reduced}")

R-squared score for the model with reduced features: 0.7096413034051532


Conclusion: Without RFE (Recursive Feature Elemination), the model was performing better. Also, I observed decreasing Test size or increasing training data the performance of model was improved.

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.